### DISCUSSION TITLE
How to use POST, PUT, PATCH, DELETE methods using fetch?

### DISCUSSION THREAD

**Discussion by mk0y** [vercel/next.js#50836] [Help]
### Summary

Hi,

I'm trying new `app` dir concept with Next.js. In particular I'm focused on server components.

While reading [this article](https://nextjs.org/docs/app/building-your-application/data-fetching/fetching) I couldn't figure out how to use anything besides GET requests? Sure, listing items is working, but how do I add new items?

I tried revalidation, but I'm not really 100% following this concept either. If I use POST requests, will Next.js re-fetch items in the server component and re-render? The comment I left [here](https://github.com/vercel/next.js/issues/50606#issuecomment-1578135136) describes this as well.

How do we have a button "Add new item" just like in client components and on onClick have an item added to the DOM re-rendered without manually refreshing the page?

My package.json:

```json
{
  "name": "next-example",
  "version": "0.1.0",
  "private": true,
  "scripts": {
    "dev": "next dev",
    "build": "next build",
    "start": "next start",
    "lint": "next lint"
  },
  "dependencies": {
    "eslint": "8.42.0",
    "eslint-config-next": "13.4.4",
    "next": "13.4.4",
    "react": "18.2.0",
    "react-dom": "18.2.0"
  }
}

```

### Additional information

```js
Results from `next info`:

    Operating System:
      Platform: darwin
      Arch: x64
      Version: Darwin Kernel Version 21.5.0: Tue Apr 26 21:08:22 PDT 2022; root:xnu-8020.121.3~4/RELEASE_X86_64
    Binaries:
      Node: 16.16.0
      npm: 8.11.0
      Yarn: N/A
      pnpm: N/A
    Relevant packages:
      next: 13.4.4
      eslint-config-next: 13.4.4
      react: 18.2.0
      react-dom: 18.2.0
      typescript: 5.1.3
```

**Comment by ciruz** [comment]
https://nextjs.org/docs/app/building-your-application/data-fetching/revalidating#on-demand-revalidation


> [On-demand Revalidation](https://nextjs.org/docs/app/building-your-application/data-fetching/revalidating#on-demand-revalidation)
If you set a revalidate time of 60, all visitors will see the same generated version of your site for one minute. The only way to invalidate the cache is if someone visits the page after the minute has passed.

> The Next.js App Router supports revalidating content on-demand based on a route or cache tag. This allows you to manually purge the Next.js cache for specific fetches, making it easier to update your site when:

> Content from your headless CMS is created or updated.
> Ecommerce metadata changes (price, description, category, reviews, etc).

You can make a regular form, as a `Client Component`, to `POST` your data to an API endpoint with `fetch`.

If you want to work with Real Time Data, you need something else like `SWR` or `React Query`.


PS: to give you an idea:

```tsx
//utils/fetch.tsx
export async function postSomething(data: DataType, params: any) {
  var myHeaders = new Headers();
  myHeaders.append("Content-Type", "application/json");
  // add more header informations if needed...

  var requestOptions = {
    method: "POST",
    headers: myHeaders,
    body: JSON.stringify(params),
  };

  return await fetch(
    process.env.SOME_API_SERVER + `/some-url/${data.id}/update`,
    requestOptions
  )
    .then((res) => {
      return res.json();
    })
    .then((json) => json);
}


//components/form.tsx
"use client"

export default function SomeForm() {
  const submit = async () => {
    postSomething(data, postData).then((json) => {
      if (json.success === false) {
        //setError(true); // oh no
      } else {
        router.push(`/success`); // oh yeah
      }
    });
  };
  
  return (
    <>
      ....
      <button onClick={() => { submit(); }}>submit</button>
    </>
  );
}
```

**Comment by mk0y** [comment]
Thanks @ciruz 

I solved it using actions. Similar to [this](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions#custom-invocation-without-starttransition) example. Then inside this function i call `revalidateTag`.

But you need experimental option in 'next.config.js' https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions#convention
